### PR TITLE
Calculator: Fixes error with percentages

### PR DESCRIPTION
**Pull request by Lord-Lavios**
## Description of new Instant Answer, or changes
Fixes queries involving calculating mod, percentage addition with decimals and percentage subtraction.
eg. 50 - 965% - ([DuckDuckGo](https://duckduckgo.com/?q=50+-+965%25&atb=v60-1&ia=calculator),  [Google](https://encrypted.google.com/search?hl=en&q=calculator#newwindow=1&hl=en&q=50+-+965%25)), (DuckDuckGo - 13%5 is 13% * 5 instead of 13 modulo 5, [Google](https://encrypted.google.com/search?hl=en&q=13%255), [WolframAlpha](http://www.wolframalpha.com/input/?i=13%255)), 50 + 5.55555% - ([DuckDuckGo](https://duckduckgo.com/?q=50+%2B+5.55555%25&atb=v60-1&ia=calculator),  [Google](https://encrypted.google.com/search?hl=en&q=calculator#newwindow=1&hl=en&q=50+%2B+5.55555%25)).

## Related Issues and Discussions
None

## People to notify
@pjhampton @moollaza 

## Testing & Review
To be completed by Community Leader (or DDG Staff) when reviewing Pull Request

**Pull Request**
- [ ] Title follows correct format (Specifies Instant Answer + Purpose)
- [ ] Description contains a valid Instant Answer Page Link (e.g. https://duck.co/ia/view/my_ia)

**Instant Answer Page** (for new Instant Answers)
- [ ] Instant Answer page is correctly filled out and contains:
    - The **Title** is appropriately named and formatted
    - The **IA topics** are present and appropriate
    - The **Description** is clear and coherent
    - **Source Name** exists if applicable
    - All **Example Queries** trigger on [Beta](https://beta.duckduckgo.com/)

**Code**
- [ ] Adheres to the [DuckDuckGo Style Guide](https://docs.duckduckhack.com/resources/code-style-guide.html)
- [ ] Behaviour is appropriately tested. If improvement, tests are adequately extended.
- [ ] There is no unnecessary files in place (such as editor config files)
- [ ] There is no API keys / secrets present
- [ ] Tests are passing (run `$ duckpan test <goodie_id>`)
    - Tester should report any failures

**Ready to merge?**

- [ ] Has this IA been deployed to and tested on [beta.duckduckgo.com](https://beta.duckduckgo.com/)?
- [ ] For larger commits, has this been approved be more than one community member?
- [ ] The number of reviews is appropriate for this type of PR
- [ ] The commit message is clear, coherent and fitting

**Pull Request Review Guidelines**: https://docs.duckduckhack.com/programming-mission/pr-review.html

---

Instant Answer Page: https://duck.co/ia/view/calculator

### CODE REVIEW

**Comment by daxtheduck** [comment]
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

File: share/goodie/calculator/calculator.js [hunk 1]
```diff
@@ -255,28 +255,16 @@ DDH.calculator = DDH.calculator || {};
...
-                return "* " + base + "." + (number < 10 ? "0" + percentage : percentage );
-            } else {
-                base += number / 100;
-                remainder = number % 100;
-                return "* " + base + "." + remainder;
-            }
+        // addSubPercentage: takes a percentage expression and rewrites it.
+        // eg. 10 + 10% --> 10 * (1 + (10/100)), 44 - 100% --> 44 * (1 - (100/100))
+        addSubPercentage: function( _expression, _operand, number, num_decimal ) {
+            return "* (1 " + _operand + "(" + number + "/" + 100 + "))";
```
Comment (pjhampton) [inline share/goodie/calculator/calculator.js]: What about 10 + 400%? You keep the base at 1?
Also, please only use `_vars` if it's unused

Comment (Lord-Lavios) [inline share/goodie/calculator/calculator.js]: > Also, please only use _vars if it's unused

I didn't know that. I just copied the parameters from the code. I'll keep that in mind.

> What about 10 + 400%? You keep the base at 1?

Yes. As I understand, we're calculating the percentages without knowing what's it being applied to. So, one way to do that was `x * (1 + y%)` which is same as `(x + xy%)`.

File: share/goodie/calculator/calculator.js [hunk 2]
```diff
@@ -87,9 +87,9 @@ DDH.calculator = DDH.calculator || {};
 
         var expression = expression
             // 1. handles +/- percentages
-            .replace(/(\+) (\d+(\.\d{1,2})?)%/g, PercentageNormalizer.addPercentage)
-            .replace(/(\d+(\.\d{1,2})?) - (\d+(\.\d{1,2})?)%/g, PercentageNormalizer.subtractPercentage)
-            .replace(/(\d+(\.\d{1,2})?)%/g, PercentageNormalizer.soloPercentage)
+            .replace(/(\+|-) (\d+(\.\d+)?)%/g, PercentageNormalizer.addSubPercentage)
+            .replace(/(\d+(\.\d+)?)%(\d+(\.\d+)?)/g, PercentageNormalizer.modPercentage)
```
Comment (moollaza) [inline share/goodie/calculator/calculator.js]: I think we want to allow for spaces here?

**Comment by pjhampton** [comment]
Yeah, I agree. This is awesome work. But it's a game of jenga without the unit tests for the js. Let's approach with caution.

**Comment by mintsoft** [comment]
So is this dead in the water given the changes to calculator since? It definitely needs a rebase, I imagine once that's done (if it's achievable) then it'll be very difficult to resolve @pjhampton

**Comment by pjhampton** [comment]
We can close this for now. In the coming days we can decide on the future of these interactive IAs 👍

**Pull request closed without merging**